PV: reorg bug fix (issue #566)

ptschip opened the pull request:
This takes care of the occasional issue we see on travis where the chain may not connect properly if we receive new blocks in the middle of  a re-org.

Several extended tests added also for the 4 and 5 block attack scenarios.

ptschip committed:
Add 4 block attack scenarios to PV python testing

Also add the needed option -pvtest which slows down the checking
of inputs by putting a sleep time of 1 second for each input to
check.  This allows up to simulate very long to validate blocks
used in the attack scenarios.

ptschip committed:
Check for reorgs in progress when new blocks are beginning validation.

Remove unnecessary code

ptschip committed:
Move block relay to ActivateBestChainStep

In PV we can connect blocks beyone pindexMostWork if other
blocks arrive just after we start validation.  Therefore we
need to relay block inventory based on the new pindexMostWork
after all the blocks have been connected.  If we instead
return to ActivateBestChain we will relay inventory based on the
old pindexMostWork and may end up not relaying a block.

ptschip committed:
Indent section

No changes to logic, just indentation and comment changes.

ptschip committed:
Add extended test for multiple block validation

Mine  blocks from different nodes at the same time repeatedly
for 100 iterations.

ptschip committed:
Add extended reorg tests for PV

There are many different reorg possibilities that are tested
in this addition to the python tests.

ptschip committed:
Fix edge case with multi reorgs

Handle a reorg that is not yet complete but another reorg
is initiated.

ptschip commented:
taking out of WIP

ptschip committed:
Run clang-format on main.cpp

gandrewstone commented on src/main.cpp:
```diff
@@ -2382,7 +2382,7 @@ bool ConnectBlock(const CBlock &block,
     bool fScriptChecks = true;
     if (pindexBestHeader)
         fScriptChecks = !fCheckpointsEnabled || block.nTime > timeBarrier ||
-                        pindex->nHeight > pindexBestHeader->nHeight - (144 * checkScriptDays.value);
```
Pedantic, but might affect regtest:
doing it this way means you could roll under 0 to a large number and then cast that to an integer.  If datatypes match this cast would = the same negative number.  But if the cast to int also changes the size of the data, then other things could happen.

So it would be better to cast before subtracting:
 (((int)pindexBestHeader->nHeight) - (144 * checkScriptDays.value));

ptschip replied:
I'll change pindex->nHeight to a uint32...it's rather odd that pindex->nHeights are int's and pindexBestHeader->nheights are uints...i think we could change that in the future but it's for another PR...I just threw this in here because of a nuisance compiler warning anyway...i don't want to spend a lot of time on it.

gandrewstone commented on src/main.cpp:
```diff
@@ -3344,6 +3346,15 @@ static bool ActivateBestChainStep(CValidationState &state,
+            // Check if the best chain has changed while we were disconnecting or processing blocks.
+            // If so then we need to return and continue processing the newer chain.
+            pindexNewMostWork = FindMostWorkChain();
+            if (pindexNewMostWork->nChainWork > pindexMostWork->nChainWork)
```
FindMostWorkChain() can return NULL

ptschip replied:
good one...fixed.

gandrewstone commented on src/main.cpp:
```diff
@@ -3446,131 +3500,104 @@ bool ActivateBestChain(CValidationState &state, const CChainParams &chainparams,
-            // Whether we have anything to do at all.
-            if (pindexMostWork == NULL)
+        // Whether we have anything to do at all.
+        if (pindexMostWork == NULL)
```
In the new code you use pindexMostWork before checking it for NULL.  Here, someone clearly thought that NULL was possible.  Perhaps this null check and return should be moved above the if (fParallel) {}?

ptschip replied:
thanks i''ll change that

gandrewstone commented on src/main.cpp:
```diff
@@ -3446,131 +3500,104 @@ bool ActivateBestChain(CValidationState &state, const CChainParams &chainparams,
-    } while (pindexMostWork > chainActive.Tip());
+
+        if (!ActivateBestChainStep(state, chainparams, pindexMostWork,
+                pblock && pblock->GetHash() == pindexMostWork->GetBlockHash() ? pblock : NULL, fParallel))
```
while some argue that basic order of operations should be known, I think that it would improve code clarity to add parens in this case since the relative order of ? vs && is pretty esoteric.  In other words, is this statement:
(pblock && pblock->GetHash() == pindexMostWork->GetBlockHash()) ? pblock : NULL
or
pblock && (pblock->GetHash() == pindexMostWork->GetBlockHash()) ? pblock : NULL)

ptschip replied:
done.

ptschip replied:
This actually failed to compile...has trouble with the casting because of 'const CBlock*'...but i did put additional parens around the whole line there , and also parens around (pblock) ... at least I think the code is much clearer although the parens around (pblock) appear a little unnecesary.

so it looks like...

((pblock) && pblock->GetHash() == pindexMostWork->GetBlockHash()) ? pblock : NULL)

ptschip committed:
return if pindexBestHeader is NULL

FindMostWorkChain() can return an NULL.

Also updated fix for compiler warning. Casting pindex->nHeight to
uint32_t.  TODO: why is pindex->nHeight an int, it really should
be uint anway.

Add Parenthesis for code clarity.
